### PR TITLE
Add PCA json endpoint

### PR DESCRIPTION
**Pull request by subdavis**
Returns `{x: Number, y: Number, label: String}`

### CODE REVIEW

**Comment by subdavis** [comment]
This is just a small change that 

* adds PCA, 
* adds constants to reference column types rather than raw strings, 
* and changes `metabolite` to `data` because I don't think anything on the backend is specialiazed for the domain of metabolomics yet.  We can also process binned NMR spectra.

**Comment by jbeezley** [comment]
I'll probably rebase this on top of my new PR.  Some of the logic here can be simplified.  Once I'm done with this, you can take a crack at fixing the indexing

**Comment by subdavis** [comment]
Sounds good to me.  Thanks.

**Comment by jbeezley** [comment]
@subdavis PTAL at my changes.

File: metabulo/models.py [hunk 1]
```diff
@@ -29,8 +30,9 @@
 })
 db = SQLAlchemy(metadata=metadata)
 
-TABLE_COLUMN_TYPES = ['key', 'metadata', 'metabolite', 'masked']
-TABLE_ROW_TYPES = ['header', 'metadata', 'sample', 'masked']
+TableTypes = namedtuple('TableTypes', ['INDEX', 'METADATA', 'DATA', 'MASK'])
+TABLE_COLUMN_TYPES = TableTypes('key', 'metadata', 'metabolite', 'masked')
+TABLE_ROW_TYPES = TableTypes('header', 'metadata', 'sample', 'masked')
```
Comment (subdavis) [inline metabulo/models.py]: if `len(column types) != len(row types)` this is easy to fix right?  We can just break the named tuples into row and col.

Comment (subdavis) [reply metabulo/models.py]: Could we trade `metabolite` and `metabolites` as strings and variable names for `measurement` and `measurements`?

I can do this in a separate PR, since it doesn't really affect this one.

File: metabulo/plot.py [hunk 2]
```diff
@@ -17,3 +18,17 @@ def make_box_plot(frame):
     output = io.BytesIO()
     FigureCanvas(fig1).print_png(output)
     return output.getvalue()
+
+
+def pca(metabolites, metadata, x_component_index, y_component_index):
+    n_components = max(x_component_index, y_component_index) + 1
```
Comment (subdavis) [inline metabulo/plot.py]: is this the same as `n_components == min(n_samples, n_features)`  ?

File: metabulo/plot.py [hunk 3]
```diff
@@ -17,3 +18,17 @@ def make_box_plot(frame):
...
+    components = PCA(n_components=n_components).fit_transform(metabolites)
+    samples = []
+    for i in range(metabolites.shape[0]):
+        samples.append({
+            'x': components[i][x_component_index],
+            'y': components[i][y_component_index],
+            'key': metabolites.index[i],
+            'labels': metadata.iloc[i].to_dict()
+        })
+    return samples
```
Comment (subdavis) [inline metabulo/plot.py]: What's your gut on writing a Marshmallow schema here to load this into?  I thought about it, but it seemed like overkill.

File: metabulo/models.py (hunk 1 above)
Comment (jbeezley) [reply metabulo/models.py]: Yes and sure.

File: metabulo/plot.py (hunk 2 above)
Comment (jbeezley) [reply metabulo/plot.py]: No.  This is just more efficient because it only computes the minimal number of singular vectors (SVD is rather slow especially if we get into hundreds of metabolites).

With the code you wrote, it was always returning the largest principal component vs. the second largest.  In that case, you only needed the first two components (the two largest singular vectors).  I modified the endpoint to scatter plot the samples projected onto arbitrary components in each axis.

BTW if we ever do anything like the PCA overview plot from metaboanalyst, it will be cheaper just to return all combinations in one call returning a (ncomponents x ncomponents x nsamples) array.

File: metabulo/views.py [hunk 4]
```diff
@@ -177,6 +177,17 @@ def get_box_plot(csv_id):
     return Response(fig, mimetype='image/png')
 
 
+@csv_bp.route('/csv/<uuid:csv_id>/plot/pca', methods=['GET'])
+def get_pca_plot(csv_id):
+    csv_file = CSVFile.query.get_or_404(csv_id)
+    x_component_index = int(request.args.get('x_component_index', 0))
+    y_component_index = int(request.args.get('y_component_index', 1))
```
Comment (subdavis) [inline metabulo/views.py]: why would I want to change these in my request?  to flip the axis?

File: metabulo/plot.py (hunk 3 above)
Comment (jbeezley) [reply metabulo/plot.py]: Seems overkill to me as well.  We could try webargs for this sort of thing.

File: metabulo/views.py (hunk 4 above)
Comment (jbeezley) [reply metabulo/views.py]: To choose which principal component is shown on each axis.  I think I remember Metaboanalyst having a feature like that.

Comment (subdavis) [reply metabulo/views.py]: Edit: nevermind

File: metabulo/plot.py (hunk 2 above)
Comment (subdavis) [reply metabulo/plot.py]: :+1:
